Enable trim analysis and fix warnings

Rob-Hague opened the pull request:
While the library is not really so modern as to think about trimming and AOT, with the default behaviour to [trim all assemblies](https://learn.microsoft.com/en-us/dotnet/core/compatibility/deployment/7.0/trim-all-assemblies) when trimming an application we have seen issues such as #1167, and I believe https://github.com/sshnet/SSH.NET/issues/859#issuecomment-1764623253 is related.

Fortunately it does not require many changes.

Rob-Hague committed:
Enable trim analysis and fix warnings

Rob-Hague commented on src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs:
```diff
@@ -41,13 +41,20 @@ public RsaDigitalSignature(RsaKey rsaKey, HashAlgorithmName hashAlgorithmName)
+            hash.AppendData(input);
+            return hash.GetHashAndReset();
+#else
+            using var hash = CryptoConfig.CreateFromName(_hashAlgorithmName.Name) as HashAlgorithm
```
`CryptoConfig` is a somewhat legacy API, and `CreateFromName` is annotated with `[RequiresUnreferencedCode]`. I'm not sure which of these points is the root cause for https://github.com/sshnet/SSH.NET/issues/859#issuecomment-1764623253 (needs more details), but either way we can just use `IncrementalHash` instead.

WojciechNagorski commented on src/Renci.SshNet/Renci.SshNet.csproj:
```diff
@@ -5,19 +5,23 @@
 	<TargetFrameworks>net462;netstandard2.0;netstandard2.1;net6.0;net7.0</TargetFrameworks>
   </PropertyGroup>
 
+  <PropertyGroup Condition=" $([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net6.0')) ">
+    <IsTrimmable>true</IsTrimmable>
+  </PropertyGroup>
```
I don't think we need to enable this feature directly in the library, right?

Rob-Hague replied:
According to https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming (if I have understood correctly) is it a library-level flag to declare that the library is compatible with trimming. It also enables the analysers.

So I think we either need `IsTrimmable` or `EnableTrimAnalyzer`, and because this PR fixes all the trimming warnings, then we can have `IsTrimmable`.

WojciechNagorski replied:
This is good answer :)

drieseng replied:
@Rob-Hague, do you happen to know how exactly it "marks the assembly as "trimmable"? Do we also mark **SshNet.Security.Cryptography** as trimmable? I would've expect a warning because we have a dependency that is marked trimmable.

Rob-Hague replied:
Good point, no, I don't know how it works with references. I have changed it just to `EnableTrimAnalyzer` until someone more knowledgeable comes along.

WojciechNagorski commented on src/Renci.SshNet/Renci.SshNet.csproj:
```diff
@@ -5,19 +5,23 @@
   </PropertyGroup>
 
-  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' or '$(TargetFramework)' == 'netstandard2.1' or '$(TargetFramework)' == 'net6.0' or '$(TargetFramework)' == 'net7.0' ">
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' or $([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'netstandard2.1')) ">
```
👍 Good idea

WojciechNagorski commented on src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs:
```diff
@@ -41,13 +41,20 @@ public RsaDigitalSignature(RsaKey rsaKey, HashAlgorithmName hashAlgorithmName)
         {
-            return _hash.ComputeHash(input);
+#if !NET462
+            using var hash = IncrementalHash.CreateHash(_hashAlgorithmName);
+            hash.AppendData(input);
+            return hash.GetHashAndReset();
```
Overall this PR looks good, but I need to understand this change before I could be able to merge this.

Rob-Hague replied:
Sure. `IncrementalHash` is a more modern and slightly lighter-weight method of hashing than `HashAlgorithm`. And it doesn't have trimming problems like `CryptoConfig.CreateFromName` does.

drieseng commented on src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs:
```diff
@@ -41,13 +41,20 @@ public RsaDigitalSignature(RsaKey rsaKey, HashAlgorithmName hashAlgorithmName)
+#if !NET462
+            using var hash = IncrementalHash.CreateHash(_hashAlgorithmName);
+            hash.AppendData(input);
+            return hash.GetHashAndReset();
```
Use `GetCurrentHash()` instead, unless we re-use the hash. See earlier remark.

Rob-Hague replied:
If I read the source correctly, it actually seems like `GetCurrentHash` does more work (by duplicating the underlying hash instance)
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderCng.cs

drieseng commented on src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs:
```diff
@@ -41,13 +41,20 @@ public RsaDigitalSignature(RsaKey rsaKey, HashAlgorithmName hashAlgorithmName)
         {
-            return _hash.ComputeHash(input);
+#if !NET462
+            using var hash = IncrementalHash.CreateHash(_hashAlgorithmName);
```
Would we benefit (performance wise) from creating the hash in the ctor and reusing it here.
Please create a small benchmark to help us decide.

Rob-Hague replied:
In addition to my other comment: this is a correctness fix, not a performance improvement. I would rather work on other things than to run a benchmark for such a change. Sorry for being blunt.

WojciechNagorski replied:
So maybe let's move the hash creating to the constructor. What do you think? this is the last change and I will be able to merge it.

Rob-Hague replied:
Makes sense. Done

Rob-Hague committed:
Use EnableTrimAnalyzer instead of IsTrimmable

I don't know how IsTrimmable works with references
(i.e. to SshNet.Security.Cryptography)

Rob-Hague committed:
Merge remote-tracking branch 'upstream/develop' into trimming

Rob-Hague committed:
Merge remote-tracking branch 'upstream/develop' into trimming

Rob-Hague committed:
Add additional aot/trimming related analyzers

WojciechNagorski commented:
I will review this next week.

WojciechNagorski committed:
Merge branch 'develop' into trimming

Rob-Hague committed:
Merge remote-tracking branch 'upstream/develop' into trimming

WojciechNagorski committed:
Merge branch 'develop' into trimming

Rob-Hague committed:
Initialise the hash instance in the constructor

WojciechNagorski reviewed (approved):
Now LGTM!

WojciechNagorski commented:
This issue has been fixed in the 2024.0.0 version.
